### PR TITLE
fix 885

### PR DESCRIPTION
**Pull request by hrgdavor**
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Does your submission pass tests?

fix for #885

@z3dev I have added a test case and removed throw from appndPoints. The reason is that closed path only means as a marker for those using it to connect last and first point later on ... like when rendering. So throwing err if somebody appnds to a closed path seems unnecessary.

I might be to quick on this judgement, but at elast we have a test case here as starting point.

### CODE REVIEW

**Comment by hrgdavor** [comment]
I will  revert the initial change and improve svg deserializer so it does not use append. Append is used that creates new geometry every time.
This is inefficient and could be avoided if deserializer collects points first and creates geometry in the end. and it would also prevent this error.

**Comment by z3dev** [comment]
@hrgdavor thanks for quick response.

as you are probably understanding, there are lots of issues, I.e. lots of bugs. It’s important to understand and localize issues within a specific scope. Sometimes those will be issues in the modeling library and sometimes there are issues in the IO libraries.

I think you are on the correct track here. The SVG library needs improvements to create usable paths or geometry.

please continue to improve the SVG IO library. I think this PR can be closed.

**Comment by hrgdavor** [comment]
@z3dev please review. This is a second attempt at fixing this.

The problem is that the test sample svg is malformed and now error is thrown
`Malformed svg path at 2:445. Path closed itself with command #29: Q2.12 -3.54 1.84 -3.17`

The problem is that the mentioned command adds last point that is the same as first point in the path, so the appendPoints closes it. So I have added a check for such case and throw an error with the location of the problematic svg.

**Comment by hrgdavor** [comment]
added parameter as requested by user in #885 

pathSelfClosed: 'error' - default throw an error
pathSelfClosed: 'trim' - ignore it
pathSelfClosed: 'split' - add an additional path.

default is throwing error that also says this option is available:
`Malformed svg path at 2:445. Path closed itself with command #29: "Q2.12 -3.54 1.84 -3.17". to avoid this error use pathSelfClosed:'split' or pathSelfClosed:'trim' option`

File: packages/io/svg-deserializer/svgElementHelpers.js [hunk 1]
```diff
@@ -108,7 +108,7 @@ const svgSvg = function (element, { customPxPmm }) {
   if ('HEIGHT' in element) { obj.height = element.HEIGHT }
   if ('VIEWBOX' in element) {
     const list = element.VIEWBOX.trim()
-    const exp = new RegExp('([\\d\\.\\-]+)[\\s,]+([\\d\\.\\-]+)[\\s,]+([\\d\\.\\-]+)[\\s,]+([\\d\\.\\-]+)', 'i')
+    const exp = /([\d.-]+)[\s,]+([\d.-]+)[\s,]+([\d.-]+)[\s,]+([\d.-]+)/i
```
Comment (z3dev) [inline packages/io/svg-deserializer/svgElementHelpers.js:111]: Why make these changes to the regular expressions? saving space?

Actually, these expressions should be created once at the file level, and reused. That would improve performance.

**Review by z3dev — CHANGES_REQUESTED** [review]
A few comments and requests.

In general, these changes look good. Some extra manual testing should be done with the 'svg samples'.

File: packages/io/svg-deserializer/index.js [hunk 2]
```diff
@@ -39,7 +39,7 @@ const shapesMapJscad = require('./shapesMapJscad')
  * @param {integer} [options.segments] - number of segments for rounded shapes
  * @param {string} [options.target] - target 2D geometry; geom2 or path2
  * @param {string} [options.version='0.0.0'] - version number to add to the metadata
- * @param {string} input - SVG data
+ * @param {string} [options.pathSelfClosed='error'] - [error||trim||split] if path self-closes with one of commands without stop command right after * @param {string} input - SVG data
```
[the remark this answers is not in the document]
Comment (hrgdavor) [reply packages/io/svg-deserializer/index.js:42]: sry, not much experience with docstrings, could you please write in comment what it should look like

File: packages/io/svg-deserializer/svgElementHelpers.js (hunk 1 above)
Comment (hrgdavor) [reply packages/io/svg-deserializer/svgElementHelpers.js:111]: standardx complained that new RegExp should not be used

File: packages/io/svg-deserializer/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/io/svg-deserializer/index.js:42]: just split the line, as the 'input' parameter is part of that long line.

File: packages/io/svg-deserializer/svgElementHelpers.js (hunk 1 above)
Comment (hrgdavor) [reply packages/io/svg-deserializer/svgElementHelpers.js:111]: reusing regex with global flag can be dangerous if not careful
https://medium.com/codesnips/js-careful-when-reusing-regex-636b92c6bf07
https://stackoverflow.com/questions/1520800/why-does-a-regexp-with-global-flag-give-wrong-results

but it looks those 2 are not global, there is no /g flag so we should be ok

File: packages/io/svg-deserializer/shapesMapGeometry.js [hunk 3]
```diff
@@ -215,182 +226,189 @@ const expandPath = (obj, svgUnitsPmm, svgUnitsX, svgUnitsY, svgUnitsV, svgGroups
...
         }
         break
       case 'T': // absolute quadratic Bézier shorthand
-        while (pts.length >= 2) {
-          cx = parseFloat(pts.shift())
-          cy = parseFloat(pts.shift())
+        while (pts.length >= i + 2) {
+          cx = parseFloat(pts[i++])
+          cy = parseFloat(pts[i++])
           paths[pathName] = geometries.path2.appendBezier({ segments, controlPoints: [[svg2cagX(qx, svgUnitsPmm), svg2cagY(qy, svgUnitsPmm)], [svg2cagX(qx, svgUnitsPmm), svg2cagY(qy, svgUnitsPmm)], [svg2cagX(cx, svgUnitsPmm), svg2cagY(cy, svgUnitsPmm)]] }, paths[pathName])
```
Comment (z3dev) [inline packages/io/svg-deserializer/shapesMapGeometry.js:340]: should ensurePath() be called here?

**Review by z3dev — COMMENTED** [review]
Left a comment as not sure if a change is required.

File: packages/io/svg-deserializer/shapesMapGeometry.js (hunk 3 above)
Comment (hrgdavor) [reply packages/io/svg-deserializer/shapesMapGeometry.js:340]: yes, nice catch. Thanks :)

**Comment by hrgdavor** [comment]
Fixed the missing ensurePath() . I thnik this is ready now. @z3dev pls review

**Review by z3dev — APPROVED** [review]
OK. This looks pretty good. Thanks, @hrgdavor 

Please test well with some additional SVG files.

**Comment by hrgdavor** [comment]
I have tested with few icons from iconmonstr
image
image
image

seems to be working well

**Pull request merged**